core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -49,7 +50,8 @@
 public abstract class FilterJoinRule extends RelOptRule {
   /** Predicate that always returns true. With this predicate, every filter
    * will be pushed into the ON clause. */
-  public static final Predicate TRUE_PREDICATE = (join, joinType, exp) -> true;
+  public static final Predicate TRUE_PREDICATE = (join, joinType, exp) ->
+      join.getConvention() != EnumerableConvention.INSTANCE;
```

Comment:
This will be changed back once [CALCITE-2973](https://issues.apache.org/jira/browse/CALCITE-2973) is done.

Reply:
Maybe add a TODO comment in the code to not forget about it?

Reply:
I don't think we should add in a Calcite convention to a logical rule,

Reply:
We thought it is logical rule. Unfortunately, this rule also applies to physical operators. And the default Enumerable operators don't support hash join with non-equi conditions, that is why we have to rule it out.

Reply:
Are we likely to see a fix to CALCITE-2973 reasonably soon? I agree that the code should not depend on EnumerableConvention, but we can live with it short-term.

Reply:
I checked with Lai Zhou, who said the fix to CALCITE-2973 can be done in a week.

Reply:
By the way, even if this is a short-term change, let's rename TRUE_PREDICATE.

It's bad form to redefine TRUE (or FALSE, 0, 1, Pi or most especially Planck's constant).

Reply:
I saw you renamed TRUE_PREDICATE to TRUE. That doesn't solve the problem. It was called TRUE_PREDICATE because it always returned true. Now it doesn't, so its name is wrong.

Maybe leave TRUE_PREDICATE as is (i.e. returning true), deprecate it, and use an inline lambda.